concurrent `git::reset_head()` can rewind new `HEAD` exported by other

## Description

https://github.com/martinvonz/jj/pull/3749#issuecomment-2130404116

Repo: https://ilya.gr/share/jj-bug-2024-05-24.tar.gz (50MB)

```
E    (@: wc1 -> wc0, HEAD: wc0)
|
D    (@: wc1, HEAD: wc0) "--ignore-working-copy"
|\
| C  (@: wc0, HEAD: wc1 -> wc0)
| |
B |  (@: wc0 -> wc1, HEAD: wc0 -> wc1) "squash"
|/
A    (@: wc0, HEAD: wc0)
```

1. The operation `B` squashes wc commit, and exports new `HEAD`.
2. Concurrently, the operation `C` observes uncommitted refs exported by `B` (but not the `HEAD` change), imports them, and commits the operation. It triggers `git::reset_head()` -> `git_repo.reset(..)`, and overwrites `HEAD` back to `wc0`.
3. These operations are merged at `D`.
4. At `E`, old `HEAD` overwritten by `C` is imported as the "new" `HEAD`.

## Expected Behavior

Perhaps, `HEAD` shouldn't be overwritten at `C`, or it should be re-exported at `D`?

I think the problem can be mitigated by:
a. not exporting unchanged `HEAD`
b. use `PreviousValue::MustExistAndMatch(..)` on exporting `HEAD` (but we'll need to deal with symbolic ref)

## Specifications

- Version: somewhere around 4478055e

--- a/lib/src/git.rs
+++ b/lib/src/git.rs
@@ -1093,24 +1093,36 @@ pub fn reset_head(mut_repo: &mut MutableRepo, wc_commit: &Commit) -> Result<(), 
     let git_repo = get_git_repo(mut_repo.store())?;
 
     let first_parent_id = &wc_commit.parent_ids()[0];
-    let first_parent = if first_parent_id != mut_repo.store().root_commit_id() {
+    let new_head_target = if first_parent_id != mut_repo.store().root_commit_id() {
         RefTarget::normal(first_parent_id.clone())
     } else {
         RefTarget::absent()
     };
 
     // If the first parent of the working copy has changed, reset the Git HEAD.
-    if mut_repo.git_head() != first_parent {
-        update_git_head(
-            &git_repo,
-            // TODO: we might want to use `PreviousValue::MustExistAndMatch` to handle concurrent
-            // modifications properly (#3754)
-            gix::refs::transaction::PreviousValue::MustExist,
-            first_parent
-                .as_normal()
-                .map(|id| gix::ObjectId::from_bytes_or_panic(id.as_bytes())),
-        )?;
-        mut_repo.set_git_head_target(first_parent);
+    let old_head_target = mut_repo.git_head();
+    if old_head_target != new_head_target {
+        let expected_ref = if let Some(id) = old_head_target.as_normal() {
+            // We have to check the actual HEAD state because we don't record a
+            // symbolic ref as such.
+            let actual_head = git_repo.head().map_err(GitExportError::from_git)?;
+            if actual_head.is_detached() {
+                let id = gix::ObjectId::from_bytes_or_panic(id.as_bytes());
+                gix::refs::transaction::PreviousValue::MustExistAndMatch(id.into())
+            } else {
+                // Just overwrite symbolic ref, which is unusual. Alternatively,
+                // maybe we can test the target ref by issuing noop edit.
+                gix::refs::transaction::PreviousValue::MustExist
+            }
+        } else {
+            // Just overwrite if unborn (or conflict), which is also unusual.
+            gix::refs::transaction::PreviousValue::MustExist
+        };
+        let new_oid = new_head_target
+            .as_normal()
+            .map(|id| gix::ObjectId::from_bytes_or_panic(id.as_bytes()));
+        update_git_head(&git_repo, expected_ref, new_oid)?;
+        mut_repo.set_git_head_target(new_head_target);
     }
 
     // If there is an ongoing operation (merge, rebase, etc.), we need to clean it

--- a/lib/tests/test_git.rs
+++ b/lib/tests/test_git.rs
@@ -39,6 +39,7 @@ use jj_lib::commit_builder::CommitBuilder;
 use jj_lib::git;
 use jj_lib::git::FailedRefExportReason;
 use jj_lib::git::GitBranchPushTargets;
+use jj_lib::git::GitExportError;
 use jj_lib::git::GitFetch;
 use jj_lib::git::GitFetchError;
 use jj_lib::git::GitImportError;
@@ -72,6 +73,7 @@ use test_case::test_case;
 use testutils::commit_transactions;
 use testutils::create_random_commit;
 use testutils::write_random_commit;
+use testutils::CommitGraphBuilder;
 use testutils::TestRepo;
 use testutils::TestRepoBackend;
 
@@ -2309,6 +2311,75 @@ fn test_reset_head_to_root() {
     assert!(tx.repo().git_head().is_absent());
     // The placeholder ref should be deleted
     assert!(git_repo.find_reference("refs/jj/root").is_err());
+}
+
+#[test]
+fn test_reset_head_detached_out_of_sync() {
+    // Create colocated workspace
+    let settings = testutils::user_settings();
+    let temp_dir = testutils::new_temp_dir();
+    let workspace_root = temp_dir.path().join("repo");
+    let git_repo = testutils::git::init(&workspace_root);
+    let (_workspace, repo) =
+        Workspace::init_external_git(&settings, &workspace_root, &workspace_root.join(".git"))
+            .unwrap();
+
+    let mut tx = repo.start_transaction();
+
+    //   4
+    //   |
+    // 2 3
+    // |/
+    // 1 5
+    // |/
+    // root
+    let mut graph_builder = CommitGraphBuilder::new(tx.repo_mut());
+    let commit1 = graph_builder.initial_commit();
+    let commit2 = graph_builder.commit_with_parents(&[&commit1]);
+    let commit3 = graph_builder.commit_with_parents(&[&commit1]);
+    let commit4 = graph_builder.commit_with_parents(&[&commit3]);
+    let commit5 = graph_builder.initial_commit();
+
+    // unborn -> commit1 (= commit2's parent)
+    git::reset_head(tx.repo_mut(), &commit2).unwrap();
+    assert_eq!(
+        tx.repo().git_head(),
+        RefTarget::normal(commit1.id().clone())
+    );
+
+    // External process updates HEAD to point to commit5
+    testutils::git::set_head_to_id(
+        &git_repo,
+        gix::ObjectId::from_bytes_or_panic(commit5.id().as_bytes()),
+    );
+
+    // {expected: commit1, actual: commit5} -> commit1 (= commit3's parent):
+    // works because the expected HEAD is unchanged.
+    git::reset_head(tx.repo_mut(), &commit3).unwrap();
+    assert_eq!(
+        tx.repo().git_head(),
+        RefTarget::normal(commit1.id().clone())
+    );
+
+    // {expected: commit1, actual: commit5} -> commit3 (= commit4's parent)
+    assert_matches!(
+        git::reset_head(tx.repo_mut(), &commit4),
+        Err(GitExportError::InternalGitError(_))
+    );
+
+    // Import the HEAD moved by external process
+    git::import_head(tx.repo_mut()).unwrap();
+    assert_eq!(
+        tx.repo().git_head(),
+        RefTarget::normal(commit5.id().clone())
+    );
+
+    // commit5 -> commit3 (= commit4's parent)
+    git::reset_head(tx.repo_mut(), &commit4).unwrap();
+    assert_eq!(
+        tx.repo().git_head(),
+        RefTarget::normal(commit3.id().clone())
+    );
 }
 
 fn get_index_state(workspace_root: &Path) -> String {

